transports/wasm-ext/src/webtransport/transport.rs

```diff
@@ -0,0 +1,100 @@
+use futures::future::FutureExt;
+use libp2p_core::address_translation;
+use libp2p_core::multiaddr::Multiaddr;
+use libp2p_core::muxing::StreamMuxerBox;
+use libp2p_core::transport::{
+    Boxed, ListenerId, Transport as TransportTrait, TransportError, TransportEvent,
+};
+use libp2p_identity::{Keypair, PeerId};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use crate::webtransport::utils::parse_multiaddr;
+use crate::webtransport::Connection;
+use crate::webtransport::Error;
+
+pub struct Config {
+    keypair: Keypair,
+}
+
+pub struct Transport {
+    config: Config,
+}
+
+impl Config {
+    pub fn new(keypair: &Keypair) -> Self {
+        Config {
+            keypair: keypair.to_owned(),
+        }
+    }
+}
+
+impl Transport {
+    pub fn new(config: Config) -> Transport {
+        Transport { config }
+    }
+}
+
+impl Transport {
+    pub fn boxed(self) -> Boxed<(PeerId, StreamMuxerBox)> {
+        self.map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
+            .boxed()
+    }
+}
+
+impl TransportTrait for Transport {
+    type Output = (PeerId, Connection);
+    type Error = Error;
+    type ListenerUpgrade = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(
+        &mut self,
+        _id: ListenerId,
+        _addr: Multiaddr,
+    ) -> Result<(), TransportError<Self::Error>> {
+        Err(Error::ListeningNotPossible.into())
+    }
+
+    fn remove_listener(&mut self, _id: ListenerId) -> bool {
+        false
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let endpoint = parse_multiaddr(&addr).map_err(|e| match e {
+            Error::InvalidMultiaddr(_) => {
+                // TODO: Log error
+                TransportError::MultiaddrNotSupported(addr)
+            }
+            e => e.into(),
+        })?;
+
+        let mut session = Connection::new(&endpoint)?;
+        let keypair = self.config.keypair.clone();
+
+        Ok(async move {
+            let peer_id = session.authenticate(&keypair, &endpoint.certhashes).await?;
+            Ok((peer_id, session))
+        }
+        .boxed())
+    }
+
+    fn dial_as_listener(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<Self::Dial, TransportError<Self::Error>> {
+        self.dial(addr)
```

Comment:
```suggestion
        Err(TransportError::MultiaddrNotSupported(addr))
```

In my eyes, on missing hole-punching support, returning `MultiaddrNotSupported` is the better approach. See also `libp2p-quic`.

https://github.com/libp2p/rust-libp2p/blob/87e863e8c9a116de79a9dd56fbacda7986bcdfe7/transports/quic/src/transport.rs#L197-L206